File: integration-tests/java-selenium/java-selenium-test/src/main/java/com/gu/fronts/integration/test/page/common/GalleryOverlay.java
package com.gu.fronts.integration.test.page.common;

import static com.gu.fronts.integration.test.fw.selenium.ByTestAttributeSelector.TEST_ATTR_NAME;

import java.util.List;

import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;

import com.gu.fronts.integration.test.fw.selenium.FindByTestAttribute;
import com.gu.fronts.integration.test.page.nwfront.NetworkFrontPage;
import com.gu.fronts.integration.test.page.util.AbstractParentPage;

public class GalleryOverlay extends AbstractParentPage {

    private static final String GALLERY_IMAGE_ITEM_ID = "gallery-image-item";
    @FindByTestAttribute(using = "gallery-grid")
    private WebElement galleryGridButton;
    @FindByTestAttribute(using = "gallery-full")
    private WebElement galleryFullButton;
    @FindByTestAttribute(using = "gallery-next")
    private WebElement galleryNextButton;
    @FindByTestAttribute(using = GALLERY_IMAGE_ITEM_ID)
    private List<WebElement> galleryImages;
    @FindByTestAttribute(using = "close-overlay")
    private WebElement closeOverlayButton;

    public GalleryOverlay(WebDriver webDriver) {
        super(webDriver);
    }

    @Override
    public GalleryOverlay isDisplayed() {
        assertExistsAndDisplayed(galleryGridButton);
        return this;
    }

    public GalleryOverlay clickGalleryGridMode() {
        assertExistsAndDisplayed(galleryGridButton);
        galleryGridButton.click();
        assertExistsAndDisplayed(galleryFullButton);
        return this;
    }

    public GalleryOverlay clickGalleryFullMode() {
        assertExistsAndDisplayed(galleryFullButton);
        galleryFullButton.click();
        assertExistsAndDisplayed(galleryGridButton);
        return this;
    }

    public GalleryOverlay clickNextGallery() {
        assertExistsAndDisplayed(galleryNextButton);
        galleryNextButton.click();
        return this;
    }

    /**
     * Only use this method when you are in full picture mode. Otherwise you would just get a random image
     */
    public WebElement getDisplayedImage() {
        WebElement displayedImage = null;
        for (WebElement image : galleryImages) {
            if (image.isDisplayed()) {
                displayedImage = image;
                break;
            }
        }
        if (displayedImage == null) {
            throw new RuntimeException("Could not find a displayed Gallery Overlay picture item with attribute ["
                    + TEST_ATTR_NAME + "=" + GALLERY_IMAGE_ITEM_ID);
        }
        return displayedImage;
    }

    public NetworkFrontPage close() {
        assertExistsAndDisplayed(closeOverlayButton);
        closeOverlayButton.click();
        return pageFactory.initPage(webDriver, NetworkFrontPage.class);
    }

}

File: integration-tests/scala-selenium/app/com/gu/fronts/integration/test/page/common/GalleryOverlay.scala
package com.gu.fronts.integration.test.page.common

import java.util.List

import scala.collection.JavaConversions.asScalaBuffer

import org.openqa.selenium.WebDriver
import org.openqa.selenium.WebElement

import com.gu.fronts.integration.test.fw.selenium.ByTestAttributeSelector.TEST_ATTR_NAME
import com.gu.fronts.integration.test.fw.selenium.FindByTestAttribute
import com.gu.fronts.integration.test.page.nwfront.NetworkFrontPage
import com.gu.fronts.integration.test.page.util.FrontsParentPage
import com.gu.fronts.integration.test.page.util.PageElementHelper.existsAndDisplayed

class GalleryOverlay(webDriver: WebDriver) extends FrontsParentPage(webDriver) {

  @FindByTestAttribute(using = "gallery-grid")
  private var galleryGridButton: WebElement = _

  @FindByTestAttribute(using = "gallery-full")
  private var galleryFullButton: WebElement = _

  @FindByTestAttribute(using = "gallery-next")
  private var galleryNextButton: WebElement = _

  @FindByTestAttribute(using = "gallery-image-item")
  private var galleryImages: List[WebElement] = _

  @FindByTestAttribute(using = "close-overlay")
  private var closeOverlayButton: WebElement = _

  override def isDisplayed(): Boolean = {
    existsAndDisplayed(galleryGridButton)
  }

  def clickGalleryGridMode(): GalleryOverlay = {
    existsAndDisplayed(galleryGridButton)
    galleryGridButton.click()
    existsAndDisplayed(galleryFullButton)
    this
  }

  def clickGalleryFullMode(): GalleryOverlay = {
    existsAndDisplayed(galleryFullButton)
    galleryFullButton.click()
    existsAndDisplayed(galleryGridButton)
    this
  }

  def clickNextGallery(): GalleryOverlay = {
    existsAndDisplayed(galleryNextButton)
    galleryNextButton.click()
    this
  }

  def getDisplayedImage(): WebElement = {
    var displayedImage: WebElement = null
    for (image <- galleryImages if image.isDisplayed) {
      displayedImage = image
      //break
    }
    if (displayedImage == null) {
      throw new RuntimeException("Could not find a displayed Gallery Overlay picture item with attribute [" + TEST_ATTR_NAME + "=gallery-image-item")
    }
    displayedImage
  }

  def close(): NetworkFrontPage = {
    existsAndDisplayed(closeOverlayButton)
    closeOverlayButton.click()
    pageFactory.initPage(webDriver, classOf[NetworkFrontPage])
  }
}